### ISSUE TITLE
#jsonValue does not recurse

### ISSUE THREAD

**Issue by beno** [apollostack/apollo-ios#29]
I run into a crash when I try to send a query with nested inputs.

    'NSInvalidArgumentException', reason: 'Invalid type in JSON write (_SwiftValue)'

I can't really tell what's going on, but my first investigations tell me that this isn't correct:

```swift
BookInput(author: AuthorInput(id: "123")).jsonValue //=> ["author": AuthorInput(id: "123")]
```

**Comment by martijnwalraven** [comment]
@beno: Thanks for reporting this! I just pushed a fix, so if you're running from master you should be able to use it immediately.

In the process, I realized we're missing some other input object features however. In particular, we need a way to differentiate between not specifying an optional field, or passing in an explicit `nil` when constructing an input object. So I'm working on some changes to the generated code, and it probably makes more sense to wait until that's finished until I publish a new release.

**Comment by beno** [comment]
Fantastic, I have switched to master just now.

I see what you mean with the `nil` setting, and I'll await your next steps...

**Comment by martijnwalraven** [comment]
@beno: I just released `apollo-codegen` 0.9.3, which generates explicit memberwise initializers for input objects to properly support optional values. I also released `apollo-ios` 0.4.2, which includes the fix in this issue and also adds some tests.
